### ISSUE TITLE
Trouble calling future within r_session$new() but not r_bg()

### ISSUE THREAD

**Issue by wlandau** [r-lib/callr#225]
This question is related to https://github.com/HenrikBengtsson/future/discussions/607, but I am not sure if the solution lies with my usage of `future` or my usage of `callr`.

Are there differences in environment variables etc. between `r_session$new()` and `r_bg()`? Is there anything I can do to make configure the R session/environment of the former be more like the latter.

I am developing a [package](https://github.com/wlandau/crew/tree/11) that requires submitting a `future` inside a separate local process, and because of a some implementation details, I would prefer to use `r_session$new()` rather than `r_bg()`. But when I try the former, I get an error:

```{r}
fun <- function() {
  plan <- future::tweak(
    future.batchtools::batchtools_sge,
    template = "sge.tmpl"
  )
  future::plan(plan) # Runs on my company's SGE cluster
  future::future("x")
}
px <- callr::r_session$new()
px$call(func = fun, args = list())
out <- px$read()
```

```{r}
cat(out$error$message)
#> callr subprocess failed: Fatal error occurred: 101.
#> Command 'qsub' produced exit code 1.
#> Output: 'Unable to run job: got no response from JSV script
#> "/opt/uge/uge-8.6.6/util/resources/jsv/verify_job".
#> Exiting.
```

This happens both in the RStudio IDE and in a terminal. In both cases, it works with `callr::r_bg()`.

```{r}
px <- callr::r_bg(func = fun, args = args)
px$get_result()
```

Here is my `sge.tmpl` file:

```bash
#!/bin/bash
#$ -cwd
#$ -j y
#$ -o <%= log.file %>
#$ -V
#$ -N <%= job.name %>
module load R/4.1.2
Rscript -e 'batchtools::doJobCollection("<%= uri %>")'
exit 0
```

**Comment by wlandau** [comment]
By the way, for the [package](https://github.com/wlandau/crew/tree/11) I mentioned, I plan to implement task queues for cloud workers, and I am gradually working up to the challenge through `callr` and `future`/`future.batchtools`. @gaborcsardi, your [task queue blog post](https://www.tidyverse.org/blog/2019/09/callr-task-q/) from 2019 was extremely helpful, and motivated the design I am using for all the queues. I credit you in the [NOTICE](https://github.com/wlandau/crew/blob/11/NOTICE) and [comments](https://github.com/wlandau/crew/blob/f864d8d56ddd8fd270046d97518ba7840c3e76c3/R/class_queue_callr.R#L1-L3), and I will add a note to the README as well.

**Comment by gaborcsardi** [comment]
> Are there differences in environment variables etc. between `r_session$new()` and `r_bg()`? 

Maybe, but you can print/save the environment variables in both and compare.

Unfortunately it is challenging for me to reproduce this, so there isn't much I can do I am afraid.

**Comment by wlandau** [comment]
That's okay, I understand.

**Comment by gaborcsardi** [comment]
@wlandau Do you think it is possible to create a self contained docker container that reproduces this?

**Comment by wlandau** [comment]
I can’t promise I will have enough time, but it is a good idea.

**Comment by wlandau** [comment]
I did figure out how to reproduce this without `future` or `batchtools`. With this `jobs.sh` script:

```bash
#!/bin/bash
#$ -cwd
#$ -j y
#$ -V
#$ -N test
sleep 5
```

This code reproduces the JSV script error:

```r
fun <- function() system2("qsub", "job.sh")
px <- callr::r_session$new()
px$call(func = fun, args = list())
out <- px$read()
out
```

and this job runs successfully.

```r
fun <- function() system2("qsub", "job.sh")
px <- callr::r_bg(fun = fun)
out <- px$get_result()
out
```

I will see if my sys admin knows what I could do to troubleshoot.

**Comment by gaborcsardi** [comment]
That's a good step. But I suspect that I would still need to set up an SGE cluster to run `qsub`.

**Comment by wlandau** [comment]
I asked my sys admin about containerizing SGE, but unfortunately he did not seem to think that was feasible. We debugged for a while, and he plans to send `strace` output to an SGE developer. I could share the trace with you by email or some other way that is not public-facing.

Also, I noticed that `r_session$new()` runs `R --no-readline --slave --no-save --no-restore`. I thought my issue might have something to do with command line flags, so I tried `R --no-readline --slave --no-save --no-restore -e 'system2("qsub", "job.sh")'` and the job submitted successfully.
